File: Lab Work/LAB3/Task2.cpp
//Name: Fatima Javaid
//Roll no: 2024-csr-010

#include <iostream>
using namespace std;

#define MAX_SIZE 5

class StackArray {
private:
    int data[MAX_SIZE];
    int top;

public:

    StackArray() : top(-1){}

    void push(int val) {
        if (top == MAX_SIZE - 1) {
            cout << "\nERROR: Stack Overflow! Cannot push " << val << "(Array is full)." << endl;
            return;
        }
        data[++top] = val;
        cout << "Pushed: " << val << endl;
    }

    int pop(){
        if (top == -1) {
            cout << "\nERROR: Stack Underflow! (Stack is empty)." << endl;
            return -1;
        }

        return data[top--];
    }

    int peek() const {
        if (top == -1) return -1;
        return data[top];
    }

    void print() const {
        cout << "StackArray (TOP->BOT): [";
        for (int i = top; i >= 0; --i) {
            cout << data[i] << (i == 0 ? "" : ", ");
        }
        cout << "]" << endl;
    }
};

int main() {
    cout << "___ Task 2 Demonstration ___" << endl;
    StackArray stack;

    stack.push(10);
    stack.push(20);
    stack.push(30);
    stack.push(40);
    stack.push(50);
    stack.push(60); // trigger overflow
    stack.print();

    

    cout << "Popped value: " << stack.pop() << endl;
    stack.print();
    return 0;
}

File: Lab Work/LAB4/Task2.py
# Name: Fatima Javaid
# Roll no: 2024-csr-010
# Section: B

# Task-2: BracketChecker (Balanced Parentheses)

class Stack:
    def __init__(self):
        """Initialize internal storage for the stack using a Python list.
        Top of the stack will be at the *end* of the list for O(1) amortized push and pop.
        """
        self._data = []  # internal list to hold items

    def push(self, item):
        """Place a new item on the top of the stack."""
        self._data.append(item)  # append adds at end (the top)

    def pop(self):
        """Remove and return the top item from the stack.
        Raises:
            IndexError: if the stack is empty (underflow).
        """
        if self.is_empty():  # guard against removing from empty stack
            raise IndexError("Stack underflow")
        return self._data.pop()  # remove and return last element

    def peek(self):
        """Return the top item without removing it.
        Raises:
            IndexError: if the stack is empty.
        """
        if self.is_empty():
            raise IndexError("Empty stack")
        return self._data[-1]  # last element is the top

    def is_empty(self):
        """Return True if the stack has no elements; otherwise False."""
        return len(self._data) == 0

    def size(self):
        """Return the current number of elements in the stack."""
        return len(self._data)

class BracketChecker:
    def __init__(self):
        """Create a new bracket checker that uses an internal Stack instance."""
        self._stack = Stack()  # composition: use a Stack inside

    def _is_open(self, ch):
        """Return True if character is an opening bracket."""
        return ch in "([{"

    def _matches(self, open_br, close_br):
        """Return True if open_br correctly matches close_br."""
        pairs = {')': '(', ']': '[', '}': '{'}  # mapping closing -> opening
        return pairs.get(close_br) == open_br

    def is_balanced(self, expr: str) -> bool:
        """Return True if expr has balanced (), [], {}; False otherwise.
        Ignores non-bracket characters.
        """
        self._stack = Stack()  # reset stack per call
        for ch in expr:
            if self._is_open(ch):
                self._stack.push(ch)  # push openings
            elif ch in ")]}":
                if self._stack.is_empty():  # unmatched closing bracket
                    return False
                top = self._stack.pop()  # get last opening
                if not self._matches(top, ch):  # mismatch pair
                    return False
            # ignore other characters
        return self._stack.is_empty()  # balanced only if nothing left


# --- Quick tests for Task-2 ---
bc = BracketChecker()
assert bc.is_balanced("{[()]}") is True     # perfectly nested
assert bc.is_balanced("([)]") is False      # crossing pairs
assert bc.is_balanced("(((())))") is True   # deep nesting
assert bc.is_balanced(")(") is False        # close before open
print("Task-2: BracketChecker OK")